### PR TITLE
Fix GetWeldToWorldDirectives for more than one model instances

### PR DESCRIPTION
**Pull request by nepfaff**
The 'return' statement ignored all but the first model instance.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370)

### CODE REVIEW

**Comment by nepfaff** [comment]
+@RussTedrake for review please

**Review by RussTedrake — CHANGES_REQUESTED** [review]
I did't write this.  @siddancha -- do you want to take a quick look?

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9N5JSQDgdU-ifTLN70:b52ftfm)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @nepfaff)

___
*[`manipulation/directives_tree.py` line 190 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/370#-O9N6ZlG5b_4VfGbw5id:-O9N6ZlG5b_4VfGbw5ie:b-z12swq) ([raw file](https://github.com/RussTedrake/manipulation/blob/c326c605abb3630b0f17f2ce7bdfaf7995de4b69/manipulation/directives_tree.py#L190)):*
> ```Python
>             if node.type == "model" and node.name in model_instance_names:
>                 directives.add(self.add_model_directives[node.name])
>                 return directives
> ```

can we add a test to test_directives_tree that breaks currently, and succeeds with this change?

**Review by nepfaff — COMMENTED** [review]
I see. I misunderstood the docstring then. Thanks for the clarifications!

My case for changing this is that in the current form, the docstring of `MakeMultibodyPlant` seems wrong:
```
model_instance_names: If specified, then only the named model instances
            will be added to the plant. Otherwise, all model instances will be added.
            `add_weld` directives connecting added model instances to each other or to
            the world are also preserved.
```
which is false if `model_instance_names=["iiwa", "wsg"]` as only the iiwa will be added.

> But in reality, most robots have their own, separate controllers, which maybe let you specify an unactuated attachment weight. So does this use case make sense?

My use case was having a single controller for both. I think we should allow both options.

> And we should explicitly enforce that child instances cannot be frozen if models are welded to each other.

From looking at `GetWeldedDescendantsAndDirectives`, I'd say it would make sense to not treat models as descendants as otherwise, there would be no use in specifying descendants. 

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9QX_Hc4Ffhwb8G5549:bknz00f)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @RussTedrake)

**Review by siddancha — COMMENTED** [review]
Agree with most of what you said.

> My case for changing this is that in the current form, the docstring of `MakeMultibodyPlant` seems wrong, which is false if `model_instance_names=["iiwa", "wsg"]` as only the iiwa will be added.

Correct, this needs to be fixed i.e. we need to support `model_instance_names=["iiwa", "wsg"]`. AND there seems to be another issue. I don't think we currently support `model_instance_names=None`, although that is the default value in `MakeMultibodyPlant()`'s constructor, and the docstring suggests that one is allowed to not specify `model_instance_names`. Currently, all calls to `MakeMultibodyPlant()` in the repo explicitly specify a `model_instance_names` list. FWIW, I don't remember `model_instance_names=None` being supported before my PR either.


> My use case was having a single controller for both. I think we should allow both options.

Agreed.

> From looking at `GetWeldedDescendantsAndDirectives`, I'd say it would make sense to not treat models as descendants as otherwise, there would be no use in specifying descendants.

Don't quite agree with this. Maybe I don't fully understand what you mean? The point of `GetWeldedDescendantsAndDirectives()` was to find models (e.g. wsg) that are welded to another model (e.g. iiwa), so that the wsg could be frozen when creating a controller plant for the iiwa. We can't do this if we don't represent the wsg as a descendant of the iiwa right?

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9QbGxi7vk0g_YUskRa:bu5hzmo)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @RussTedrake)

**Review by nepfaff — COMMENTED** [review]
> so that the wsg could be frozen when creating a controller plant for the iiwa

I was trying to say that in this case, we would specify `model_instance_names=["iiwa"]` and not `model_instance_names=["iiwa", "wsg"]`.

But if we specify `model_instance_names=["iiwa", "wsg"]`, then we don't want to freeze the `wsg`. In this case, we don't want to treat the `wsg` as a descendant but anything connected to the `wsg` or anything connected to the `iiwa` that is not the `wsg` as a descendant. 

So basically we ensure that anything in `model_instance_names` is not returned as a descendant.

Do you think there would be use cases for which this would be an issue?

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9Qsk5h-kvlh0iryvrt:b21oq9t)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @RussTedrake)

**Review by nepfaff — COMMENTED** [review]
Agree that we need to support `model\_instance\_names=None` as adding all models seems like a reasonable request.

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9QuEI14WaNASLlgKsL:b-597ih5)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @RussTedrake)

**Review by siddancha — COMMENTED** [review]
> I was trying to say that in this case, we would specify `model_instance_names=["iiwa"]` and not `model_instance_names=["iiwa", "wsg"]`. But if we specify `model_instance_names=["iiwa", "wsg"]`, then we don't want to freeze the `wsg`. In this case, we don't want to treat the `wsg` as a descendant but anything connected to the `wsg` or anything connected to the `iiwa` that is not the `wsg` as a descendant.

@nepfaff I see. Yes, that makes perfect sense to me! As I see it, these are the todo items for this PR (including some you've already completed):

- [x] Remove the return statement in the base case of `GetWeldToWorldDirectives()`.
- [ ] Rename `GetWeldToWorldDirectives()` to `GetDirectivesFromRootToModels()`.
- [ ] In `GetWeldedDescendantsAndDirectives()`, subtract `model_instance_names` from the set `descendants` right before returning.
- [ ] Now we have an issue where the directives in the tree between iiwa and wsg are repeated in `GetDirectivesFromRootToModels()` and `GetWeldedDescendantsAndDirectives()`. So we can't just extend the list of directives like we're doing here ([LINK](https://github.com/RussTedrake/manipulation/blob/dbaa6da059c6eeb9c7865ad9e8c4a1850db5cfbd/manipulation/station.py#L404)). Instead, I think it's better for `GetDirectivesFromRootToModels()` and `GetWeldedDescendantsAndDirectives()` to output sets instead of lists, take unions of sets, and convert it to a topologically sorted list right before calling `ProcessModelDirectives` [here](https://github.com/RussTedrake/manipulation/blob/dbaa6da059c6eeb9c7865ad9e8c4a1850db5cfbd/manipulation/station.py#L406), using my `TopologicallySortDirectives()` function.
- [ ] When `model_instance_names=None`, we first:
  - [ ] Assert that `add_frozen_child_instances` should be False; it doesn't make sense to freeze children if the user wants to add all models. 
  - [ ] Skip creating a `DirectivesTree` and directly pass flattened directives to `ProcessModelDirectives` bypassing [these lines](https://github.com/RussTedrake/manipulation/blob/dbaa6da059c6eeb9c7865ad9e8c4a1850db5cfbd/manipulation/station.py#L396-L404).
- [x] Add test case for `model_instance_names=["iiwa", "wsg"]`.
- [ ] Add test case for `model_instance_names=None`.

Does this plan sound good? Am I missing anything?

If this plan sounds good to you, either:
- I can implement this (happy to!). But you need to give me write access to your fork for me to push commits to your PR's branch. You can revoke my access once we're done with this PR. Then you review my changes.
- You implement this proposal and I will review it.

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9Qxfw31jnTyuFzlSVP:b-7vrh0)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @RussTedrake)

**Review by nepfaff — COMMENTED** [review]
All of this sounds great. Thank you very much for creating this detailed list! I don't think you're missing anything.

I can implement this now. Thanks for offering to do it though!

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9RCiLI0FW3FduTCDYM:b27428k)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @RussTedrake)

**Review by nepfaff — COMMENTED** [review]
Everything should be implemented now. @siddancha would you mind taking a look?

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9RMXU9EJLi3ViFdqGH:bxnukpe)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @RussTedrake)

**Review by siddancha — COMMENTED** [review]
Looks excellent!!! Thanks @nepfaff! I suggested adding another test, but otherwise looks great!
:lgtm:

> Reviewed 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9RaUil2rP6kVFeIrWR:boeu21f)* status: 2 of 3 files reviewed, 2 unresolved discussions (waiting on @nepfaff and @RussTedrake)

___
*[`manipulation/test/test_directives_tree.py` line 149 at r3](https://reviewable.io/reviews/RussTedrake/manipulation/370#-O9R_zC36SG0whey9qWR:-O9R_zC36SG0whey9qWS:b-hccpze) ([raw file](https://github.com/RussTedrake/manipulation/blob/b12dcbbea89d5906b412f4c8e8e538a90ff5e4fc/manipulation/test/test_directives_tree.py#L149)):*
> ```Python
>         self.assertTrue(plant.HasModelInstanceNamed("iiwa"))
>         self.assertFalse(plant.HasModelInstanceNamed("wsg"))
>         self.assertFalse(plant.HasModelInstanceNamed("table"))
> ```

Could you please add this test as well to `test_make_multibody_plant()`?

```python
        # Test with only "iiwa" and "add_frozen_child_instances=True"
        plant: MultibodyPlant = MakeMultibodyPlant(
            scenario, model_instance_names=["iiwa"],
            add_frozen_child_instances=True,
        )
        self.assertTrue(plant.HasModelInstanceNamed("iiwa"))
        self.assertTrue(plant.HasModelInstanceNamed("wsg"))
        self.assertFalse(plant.HasModelInstanceNamed("table"))


```

**Review by nepfaff — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370)* status: 2 of 3 files reviewed, 2 unresolved discussions (waiting on @RussTedrake and @siddancha)

___
*[`manipulation/test/test_directives_tree.py` line 149 at r3](https://reviewable.io/reviews/RussTedrake/manipulation/370#-O9R_zC36SG0whey9qWR:-O9Uomh585gP25Q5xxa_:bb7f2mf) ([raw file](https://github.com/RussTedrake/manipulation/blob/b12dcbbea89d5906b412f4c8e8e538a90ff5e4fc/manipulation/test/test_directives_tree.py#L149)):*
<details><summary><i>Previously, siddancha (Siddharth Ancha) wrote…</i></summary><blockquote>

Could you please add this test as well to `test_make_multibody_plant()`?

```python
        # Test with only "iiwa" and "add_frozen_child_instances=True"
        plant: MultibodyPlant = MakeMultibodyPlant(
            scenario, model_instance_names=["iiwa"],
            add_frozen_child_instances=True,
        )
        self.assertTrue(plant.HasModelInstanceNamed("iiwa"))
        self.assertTrue(plant.HasModelInstanceNamed("wsg"))
        self.assertFalse(plant.HasModelInstanceNamed("table"))


```
</blockquote></details>

done

**Review by siddancha — COMMENTED** [review]
:lgtm:

@RussTedrake I think we're ready to merge!

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/370#-:-O9VJW8H1banjM-Rd0V1:b-g029r2)* status: all files reviewed, 1 unresolved discussion (waiting on @RussTedrake)

**Comment by nepfaff** [comment]
Thanks @siddancha! Will merge to unblock the course PRs
